src/Codegen/InputObjectType.hack

```diff
@@ -0,0 +1,98 @@
+            ->setIsFinal()
+            ->setExtendsf('\%s', \Slack\GraphQL\Types\InputObjectType::class)
+            ->addTypeConstant(
+                $cg->codegenTypeConstant('TCoerced')->setValue('\\'.$hack_type, HackBuilderValues::literal()),
+            );
+
+        $name_const = $cg
+            ->codegenClassConstant('NAME')
+            ->setValue($this->input_type->getType(), HackBuilderValues::export());
+
+        $class->addConstant($name_const);
+
+        $ts = $this->reflection_type_alias->getResolvedTypeStructure();
+        invariant(
+            $ts['kind'] === \HH\TypeStructureKind::OF_SHAPE && !idx($ts, 'nullable', false),
+            'Input objects can only be generated from type aliases of a non-nullable shape type, got %s.',
+            TypeStructureKind::getNames()[$ts['kind']],
+        );
+
+        $class->addConstant(
+            $cg->codegenClassConstant('keyset<string> FIELD_NAMES')
+                ->setValue(Keyset\keys($ts['fields']), HackBuilderValues::export()),
+        );
+
+        // coerceFieldValues() and coerceFieldNodes()
+        $values = hb($cg)->addLine('return shape(')->indent();
+        $nodes = hb($cg)->addLine('return shape(')->indent();
+
+        foreach (($ts['fields'] as nonnull) as $field_name => $field_ts) {
+            $name_literal = \var_export($field_name, true);
+            $type = input_type(self::typeStructureToHackType($field_ts));
+            $values->addLinef('%s => %s->coerceValue($args[%s]),', $name_literal, $type, $name_literal);
+            $nodes->addLinef('%s => %s->coerceNode($args[%s], $vars),', $name_literal, $type, $name_literal);
+        }
+
+        $values->unindent()->addLine(');');
+        $nodes->unindent()->addLine(');');
+
+        return $class->addMethods(vec[
+            $cg->codegenMethod('coerceFieldValues')
+                ->setIsOverride()
+                ->addParameter('KeyedContainer<arraykey, mixed> $args')
+                ->setReturnType('this::TCoerced')
+                ->setBody($values->getCode()),
+            $cg->codegenMethod('coerceFieldNodes')
+                ->setIsOverride()
+                ->addParameterf('KeyedContainer<string, \\%s> $args', \Graphpinator\Parser\Value\Value::class)
+                ->addParameter('dict<string, mixed> $vars')
+                ->setReturnType('this::TCoerced')
+                ->setBody($nodes->getCode()),
+        ]);
+    }
+
+    private static function typeStructureToHackType<T>(TypeStructure<T> $ts): string {
+        $alias = Shapes::idx($ts, 'alias');
+        if ($alias is nonnull) {
+            return $alias;
+        }
+
+        switch ($ts['kind']) {
+            case TypeStructureKind::OF_INT:
+                return 'HH\int';
+            case TypeStructureKind::OF_STRING:
+                return 'HH\string';
+            case TypeStructureKind::OF_BOOL:
+                return 'HH\bool';
+            case TypeStructureKind::OF_VEC:
```

Comment:
Should we handle keyset and just treat that as vec here?

Reply:
This could actually get quite tricky. If we support keyset fields on input object shapes, there would no longer be 1:1 mapping between GraphQL types and Hack types (now `[Int]` could be either `vec<int>` or `keyset<int>`), which can complicate the coercion logic significantly, especially when variables are involved -- e.g. a GraphQL variable `$foo: [Int]` could be used in multiple places in the GraphQL query, one of which needs coercion to `vec` and the other to `keyset`. This means we could no longer coerce variables at the beginning of execution and then just use them, there would have to be an extra coercion step at each usage...

Reply:
It is definitely doable though, if people think it would be useful.

Reply:
Also, this would actually be pretty easy for *output* types. We can support `<<GraphQL\Field>>` on methods that return keyset/Container/... easily.

Reply:
Yea, I think the output types make more sense to me since it's fine to represent that as a vec in GraphQL